helpers/embeds.js

```diff
@@ -0,0 +1,65 @@
+import { html, nothing, render } from 'lit';
+import { ifDefined } from 'lit/directives/if-defined.js';
+import { map } from 'lit/directives/map.js';
+import { tryGetPluginByKey } from '@brightspace-ui/core/helpers/plugins.js';
+
+const embedTypeAttributeName = 'data-d2l-embed-type';
+
+export async function createEmbedPlaceholder(embedType, options) {
+	const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', embedType);
+	if (!embedRendererPlugin) return;
+
+	const placeholderData = await embedRendererPlugin.getPlaceholder(options);
+	const contents = placeholderData.contents
+		? map(Object.entries(placeholderData.contents), ([id, content]) => html`<template data-d2l-embed-template-id=${id}>${content}</template>`)
+		: nothing;
+
+	const attributes = placeholderData.attributes ? JSON.stringify(placeholderData.attributes) : undefined;
+
+	return placeholderData.inline
+		? html`
+			<span data-d2l-embed-type="${embedType}" data-d2l-embed-props="${ifDefined(attributes)}">
+				${contents}
+			</span>
+		` : html`
+			<div data-d2l-embed-type="${embedType}" data-d2l-embed-props="${ifDefined(attributes)}">
+				${contents}
+			</div>
+		`;
+}
+
+export async function renderEmbeds(node, options) {
+	const placeholders = [...node.querySelectorAll(`div[${embedTypeAttributeName}], span[${embedTypeAttributeName}]`)];
+	if (placeholders.length === 0) return;
+
+	const processPlaceholder = async placeholder => {
+		const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', placeholder.dataset.d2lEmbedType);
+		if (!embedRendererPlugin) return;
+
+		const templates = [...placeholder.querySelectorAll('template[data-d2l-embed-template-id]')];
+		const processedTemplates = await Promise.all(templates.map(async template => {
+			const templateNode = template.content.cloneNode(true);
+			await renderEmbeds(templateNode, options);
+			return html`${templateNode}`;
+		}));
+
+		const processedTemplateContents = templates.reduce((acc, template, index) => {
+			acc[template.dataset.d2lEmbedTemplateId] = processedTemplates[index];
+			return acc;
+		}, {});
+
+		let props = placeholder.dataset.d2lEmbedProps;
+		try {
+			props = JSON.parse(placeholder.dataset.d2lEmbedProps);
+		} catch (e) { /* empty */ }
+
+		return embedRendererPlugin.renderView(processedTemplateContents, props, options || {});
+	};
+
+	const embeds = await Promise.all(placeholders.map(processPlaceholder));
+	placeholders.forEach((placeholder, index) => {
+		if (!embeds[index]) return;
+		render(embeds[index], placeholder.parentNode, { renderBefore: placeholder });
+		placeholder.remove();
+	});
```

Comment:
The way this is structured now, we're converting between Lit template literals and DOM nodes a fair bit here. I'm not really sure of a great way to avoid that, since we need the template literals to interface between the plugins and the framework, but we also need the raw DOM nodes to do things like extract the template contents and query for placeholders.

This works out of the box in a Content file because we're working purely with raw HTML nodes to begin with, but for the html-block use case where we're starting with a Lit template, this maybe gets more complicated. It's possible that we'll need to render the Lit template into a document fragment in order to pass it down through the framework, then do something like ``return html`${documentFragment}`;`` in order to extract it out again.